### DISCUSSION TITLE
[@types/react]: Add brand to memoized values

### DISCUSSION THREAD

**Discussion by jeengbe** [DefinitelyTyped/DefinitelyTyped#65360] [Issues with a @types package]
Mainly exploring an idea, but often it can be useful to require a memoized value. In dependency arrays, for instance, it can be useful to only allow these memoized-branded types:

```ts
type Memoized = unique symbol;
// ...
function useCallback<T extends Function>(callback: T, deps: DependencyList): T & Memoized;
function useMemo<T>(factory: () => T, deps: DependencyList | undefined): T & Memoized;
```

With this, we could update the type
```diff
- type DependencyList = ReadonlyArray<unknown>;
+ type DependencyList = ReadonlyArray<Primitive | Memoized>;
```

This would introduce type-level safety for memoization and unnecessary re-renders.

Thoughts on this; would this break elsewhere?

[a comment by an automated account is not shown]

**Comment by Ayc0** [comment]
On some custom types I was working on some custom types, I did something like this [TypeScript Play](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQdAJlrgDbKVy4QB2q8AWQCiAgPIAuOAFcewAI5SscVAE8QAIwit6dLAA9IsODBVglArCAjAAXlkYAeACoA+OAF44AbziVkjXlYVOABtYTEAXUkyEEtrO0YyOABfOAAyOCcdAHpsuAFkFXUlAHclHix7YwgURkY4K04YGpBkGFwACxRWVjgwKFBgGGA6EzM4AAUBkCHgADcsdHc6OFW4AB9lGAGeAHMVtc2eKQ0sKAPVo6kei43pHmYCYArGW83VDS03uHVgXeeYDoxuY4rZ7E5TIsPJNprMFuhNhYrGDHMgeCoXEDIXAAKqoLBImqeZwuAAUtwIeGaUBUklJAEoPG4nAAaW7MMCoSTYfyBFQAQSgUEKDkJKIhZlQbk2Mkez3sbMZ7jcYoSJPo+kM8G4fHgUnxhOhPPwADp9QS4ih0HiLVYseMbQBhZA9dR4ADW0OccH0MCwD3QADEZPhgLwybdcC7WG7cO7JKz2VhOdycAEeEFBcKVKLQQkJYsXIqmfk8-Z1boDNBtbx+NJ8c7XR6jVQYGaG9HY560LiO024zoAWdKbglDbsARvLcSaTnrMXQA1F2KBP0ySq8vGtsCKQwZDqVhYCeidQAKxYMBJmOnrlnsmGi+XWATd2OPTXpeRaq3JuPZ4vV70GscDep4spYE8Lxkh+G6OD+O57geR4QSe574N6MoPBB8qMC417JJWWpcLWer4hOLZUu2yGTj244QYOPB+lAI5jviADKe5+lOwEOGxZJzg+rAcW0z5wGxdykgyJZsfSH4hLcwGwbxRYKWsSk-gAIsAqBgG0nQOD+bFYDAwl+vyoa8MpeFssBERAWsvHQuBkH2NBkjycBillo44mYXKUE2Z5qzqa2JpaTpekdAZoVGSZnFYOZwyWb59z+a51m3HZdAEZq1bEbq9ZYKZSieD+5rFVavZFfFDFMSxVXYIwUijlA3EORgPp6H6AaYKFjXNWcADqQwdBAu6JWGPAOGiGIsnAACSEZBaslBNS13KBUFAnAC6YJQIKuySPNm2edtu12FAdIcIdC1Km4P79S1xXDTAo3jRZU0YCpwFyapwXedFlGPWcz0jWNMATZZX3fctIWUeFuntB0L1vRDH0w6sWU8R1vr+ow6APfYA1QCj4OQ1NM14eSy2+ET62YCdwFnawe0HamQN0yD8Wk+9SWfRjp33jtLMXQA-JIzk4SdcmE2tXMiTzaN8wZRa9fD2mI50ivk1j7Wdd1+Nq6awNQNN6JzZTc2LdTy2rcTG1-XAzOs1AN3zekRttibxUq4zazOxdV2u0dHuy8TPtfXdnu-pzUARwLMt9bH8dzZpGuRYDxux+TKsuLrqwGfreME0ncumzNFvoqr1uO3b9MYH7qwB2cbMLY3TtC+dZxB27Udh098W+7cicc2XKfRwjGf9y3H25-nIE411xfRybZsqJXGJLbbscOzTzPFezWdj4PX3t83UDi6l2EvNL7nT3HJ+q2nEVI5nXvZ7P0NZTlVZGDqdbmhNhRU0gDY6VTomXegQA)

I think this could be nice to add into `@types/react`. But:
1. This is a **huge breaking change** (so we'll need to wait for at least the v19)
2. Maybe we can expose this behind another file, like `/// <reference path="force-memoized.d.ts" />` (like for the experimental types), so that those could be opt-in manually without requiring a major upgrade
